File: truncated_gaussian_runtime.R
library(tmg)
library(TruncatedNormal)
library(cdists)
library(tictoc)

library(data.table)

# ptnorm <- function(x, lb = -Inf, ub = Inf) {
#   # 1d truncated normal
#   truncdist::ptrunc(x, spec = "norm", a = lb, b = ub)
# }

n <- 1000
dimensions <- c(100, 500, 1000, 2000, 4000, 6000, 8000, 10000)


# Identity covariance
esses <- vector(mode = "list", length = length(dimensions))
names(esses) <- dimensions
samples <- vector(mode = "list", length = length(dimensions))
names(samples) <- dimensions
times <- vector(mode = "list", length = length(dimensions))
names(times) <- dimensions
dists <- vector(mode = "list", length = length(dimensions))
names(dists) <- dimensions

set.seed(1)
uni_sample <- tmvtnorm::rtmvnorm(n, 0, 1, 0)
for (d in dimensions) {
  print(paste("--- Dimension ", d, " ---"))
  mu <- rep(0, d)
  Sigma <- diag(d)
  lb <- rep(0, d)
  ub <- rep(Inf, d)
  
  # print("Botev")
  # tic()
  # botev_samples <- TruncatedNormal::rtmvnorm(n, mu, Sigma, lb, ub)
  # botev_time <- toc()
  # botev_ess <- coda::effectiveSize(t(botev_samples))
  # botev_dist <- apply(botev_samples, 2, 
  #                         function(samp) ks.test(samp, uni_sample)$statistic)
  
  print("Exact")
  tic()
  exact_samples <- cdists::rtmvn(n, Sigma, mu, lb, ub)
  exact_time <- toc()
  exact_ess <- coda::effectiveSize(exact_samples)
  exact_dist <- apply(exact_samples, 2,
                          function(samp) ks.test(samp, uni_sample)$statistic)
  
  entry_name <- as.character(d)
  
  # esses[[entry_name]] <- data.frame(botev = botev_ess, exact = exact_ess, 
  #                                   d = d, dim_idx = 1:d)
  # times[[entry_name]] <- data.frame(
  #   botev = with(botev_time, toc - tic),
  #   exact = with(exact_time, toc - tic), d = d)
  # dists[[entry_name]] <- data.frame(botev = botev_dist, exact = exact_dist, 
  #                                   d = d, dim_idx = 1:d)
  
  esses[[entry_name]] <- data.frame(exact = exact_ess, d = d, dim_idx = 1:d)
  times[[entry_name]] <- data.frame(exact = with(exact_time, toc - tic), d = d)
  dists[[entry_name]] <- data.frame(exact = exact_dist, d = d, dim_idx = 1:d)
}

save(esses, samples, times, dists, file = "experiments_2020_2_18.RData")


# visualization ----
library(ggplot2)
library(ggridges)

rhmc <- R.matlab::readMat("rhmc_2020_2_18.mat")
rhmc_samples <- 
  sapply(rhmc$samples, function(x) rbind(t(x[[1]]), ncol(x[[1]])))
rhmc_dims <- as.numeric(rhmc$dimensions)

rhmc_time_df <- 
  data.frame(variable = "rhmc", value = rhmc$times, d = dimensions)
rhmc_dist <- sapply(rhmc_samples, function(x) {
  apply(x, 2, function(samp) 
    ks.test(samp, uni_sample)$statistic)
  })
rhmc_dist_df <- mapply(function(distances, d) {
  data.frame(d = d, variable = "rhmc", value = distances)
}, rhmc_dist, rhmc_dims, SIMPLIFY=FALSE)
rhmc_dist_df <- do.call(rbind, rhmc_dist_df)

dist_df <- do.call(rbind, dists)
dist_df <- melt(data.table(dist_df[, -4]), id.vars = "d")
times_df <- melt(data.table(do.call(rbind, times)), id.vars = "d")

ggplot(dist_df, aes(x = variable, y = value)) +
  geom_violin() +
  facet_wrap(vars(d), scales = "free")
ggplot(dist_df, aes(x = dim_idx, y = botev)) +
  geom_point() +
  facet_wrap(vars(d), scales = "free")

dist_df <- do.call(rbind, dists)
dist_df <- melt(data.table(dist_df[, -3]), id.vars = "d")
all_dist_df <- rbind(dist_df, rhmc_dist_df)
ggplot(all_dist_df, 
       aes(x = d, y = value, group = interaction(variable, d), fill = variable)) +
  geom_boxplot(width = 1000) +
  labs(y = "KS distance to univariate truncated normal", x = "dimension")


ggplot(rbind(rhmc_time_df, times_df), 
       aes(x = d, y = value, color = variable)) + 
  geom_point(size = 3) + geom_line() + 
  labs(x = "Dimension", y = "Seconds") + 
  theme(legend.title = element_blank())

par(mfrow = c(5, 1))
plot()

init <- rep(1, d)
f <- diag(d)
g <- lb
tic()
pp_samples <- tmg::rtmg(n, Sigma, mu, init, f, g)
toc()

File: truncated_gaussian_runtime.m
%% positive orthant

P2 = Problem; 
P2.lb = zeros(2, 1);
% P2.ub = ones(2, 1);
P2.df = @(x) x;

iter = 1000;

tic;
plan = prepare(P2, struct('display', 1));
out = sample(plan, iter);
t = toc;

scatter(out.samples(1, :), out.samples(2, :))

%%
initSampler;

rng(1);
% iter = 100000;
% dimensions = [100, 500, 1000, 2000, 4000, 6000, 8000, 10000];
n_sizes = [1000, 10000, 100000];
len_sizes = length(n_sizes);
dimensions = [100, 500, 1000, 2000];
len_dim = length(dimensions);
times = inf(len_dim, len_sizes);
samples = cell(len_dim, len_sizes);

for i = 1:len_dim
    for j = 1:len_sizes
        d = dimensions(i);
        n = n_sizes(j);
        
        lb = zeros(d, 1);
        ub = inf(d, 1);
        mu = zeros(d, 1);
        Sigma = eye(d);
        P = make_orthant_problem(lb, ub, mu, Sigma);
        
        tic;
        plan = prepare(P);
        out = sample(plan, n);
        t = toc;

        times(i, j) = t;
        samples{i, j} = out.samples';
    end
end

save(['rhmc_', date, '.mat'], 'times', 'samples', 'dimensions');
% save('rhmc_2020_2_18.mat', 'times', 'samples', 'dimensions')

%% dense covariance
initSampler;

rng(1);
iter = 1000;
% dimensions = [100, 500, 1000, 2000, 4000, 6000, 8000, 10000];
dimensions = [100, 500, 1000, 2000, 4000, 6000];
len_dim = length(dimensions);
times = inf(len_dim, 1);
samples = cell(len_dim, 1);

for i = 1:len_dim
    d = dimensions(i);
    
    lb = .5 * ones(d, 1);
    ub = ones(d, 1);
    
    mu = zeros(d, 1);
    Sigma = .5 * eye(d) + .5 * ones(d, 1) * ones(1, d);
    P = make_orthant_problem(lb, ub, mu, Sigma);
    
    tic;
    plan = prepare(P);
    out = sample(plan, iter);
    t = toc;
    
    times(i) = t;
    samples{i} = out.samples;
end

save('rhmc_2020_3_4.mat', 'times', 'samples', 'dimensions')